pyfifotax/utils.py

```diff
@@ -0,0 +1,669 @@
+    for k, v in forex_dict.items():
+        for f in v:
+            # filter based on date of fee / taxation /etc. event
+            if f.date.year == report_year:
+                filtered_dict[k].append(f)
+    for _, v in filtered_dict.items():
+        v.sort(key=get_date)
+    return filtered_dict
+
+
+def forex_dict_to_df(forex_dict, mode):
+    assert mode.lower() in ["daily", "monthly"]
+    tmp = {
+        "Symbol": [],
+        "Comment": [],
+        "Date": [],
+        "Amount": [],
+        "Amount [EUR]": [],
+    }
+    for k, v in forex_dict.items():
+        for f in v:
+            tmp["Symbol"].append(k)
+            tmp["Comment"].append(f.comment)
+            date = f"{f.date.year}-{f.date.month:02}-{f.date.day:02}"
+            tmp["Date"].append(date)
+            amount = f"{f.amount:.2f} {f.currency}"
+            tmp["Amount"].append(amount)
+            if mode == "daily":
+                tmp["Amount [EUR]"].append(
+                    round_decimal(f.amount_eur_daily, precision="0.01")
+                )
+            else:
+                tmp["Amount [EUR]"].append(
+                    round_decimal(f.amount_eur_monthly, precision="0.01")
+                )
+
+    df = pd.DataFrame(
+        tmp, columns=["Symbol", "Comment", "Date", "Amount", "Amount [EUR]"]
+    )
+    return df
+
+
+def apply_rates_transact_dict(trans_dict, daily_rates, monthly_rates):
+    for v in trans_dict.values():
+        for f in v:
+            buy_price, sell_price = to_decimal(f.buy_price), to_decimal(f.sell_price)
+
+            # exchange rates are in 1 EUR : X FOREX
+            buy_rate_daily = get_daily_rate(daily_rates, f.buy_date, f.currency)
+            buy_rate_monthly = get_monthly_rate(monthly_rates, f.buy_date, f.currency)
+
+            sell_rate_daily = get_daily_rate(daily_rates, f.sell_date, f.currency)
+            sell_rate_monthly = get_monthly_rate(monthly_rates, f.sell_date, f.currency)
+
+            f.buy_price_eur_daily = buy_price / buy_rate_daily
+            f.buy_price_eur_monthly = buy_price / buy_rate_monthly
+            f.sell_price_eur_daily = sell_price / sell_rate_daily
+            f.sell_price_eur_monthly = sell_price / sell_rate_monthly
+            f.gain_eur_daily = f.quantity * (
+                f.sell_price_eur_daily - f.buy_price_eur_daily
+            )
+            f.gain_eur_monthly = f.quantity * (
+                f.sell_price_eur_monthly - f.buy_price_eur_monthly
+            )
+
+
+def filter_transact_dict(trans_dict, report_year, speculative_period=None):
+    filtered_dict = {k: [] for k in trans_dict.keys()}
+    for k, v in trans_dict.items():
+        for f in v:
+            # filter based on sell date
+            if f.sell_date.year == report_year:
+                if speculative_period is None:
+                    filtered_dict[k].append(f)
+                elif (f.sell_date - f.buy_date).days < speculative_period * 365:
```

Comment:
I received an exception here which I'm still debugging where it can came from. Sharing it now because probably only tomorrow I have time to continue.

```
Traceback (most recent call last):
  File "timestamps.pyx", line 1075, in pandas._libs.tslibs.timestamps._Timestamp._as_creso
  File "np_datetime.pyx", line 683, in pandas._libs.tslibs.np_datetime.convert_reso
OverflowError: result would overflow

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File ".../pycharm-community-2022.2.1/plugins/python-ce/helpers/pydev/pydevd.py", line 1546, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../pycharm-community-2022.2.1/plugins/python-ce/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File ".../pyFIFOtax/create_report.py", line 81, in <module>
    main(
  File ".../pyFIFOtax/create_report.py", line 73, in main
    report.create_excel_report(
  File ".../pyFIFOtax/pyfifotax/report_data.py", line 505, in create_excel_report
    dfs = self.consolidate_report(report_year, mode)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../pyFIFOtax/pyfifotax/report_data.py", line 435, in consolidate_report
    filtered_sold_forex = filter_transact_dict(
                          ^^^^^^^^^^^^^^^^^^^^^
  File ".../pyFIFOtax/pyfifotax/utils.py", line 598, in filter_transact_dict
    elif (f.sell_date - f.buy_date).days < speculative_period * 365:
          ~~~~~~~~~~~~^~~~~~~~~~~~
  File "timestamps.pyx", line 523, in pandas._libs.tslibs.timestamps._Timestamp.__sub__
  File "timestamps.pyx", line 1078, in pandas._libs.tslibs.timestamps._Timestamp._as_creso
pandas._libs.tslibs.np_datetime.OutOfBoundsDatetime: Cannot cast 0001-01-01 00:00:00 to unit='ns' without overflow.
python-BaseException
```

Reply:
I see, let me actually completely switch to just `datetime.date` instead of `datetime.datetime`. (1, 1, 1) technically only should be used as a unused dummy date for certain things. The datetime difference (which is internally done in ns) then indeed is likely to be out of bounds.
The other question would be why this difference is calculated to begin with. Is this a "EUR" balance by any chance?

Reply:
It is, indeed!

```python
if speculative_period is not None:
  print(f)
```

This is what this print tells me (without applying your latest patch):

```
HUF(Quantity: 63.12, Buy-Date: 2019-Jun-14)
EUR(Quantity: 190.43, Buy-Date: 1-Jan-01)
```

Reply:
added a special case for "EUR", together with not using datetime.datetime, things hopefully should be more robust now

Reply:
Thanks for the fixes, it works now!

Interestingly, in most of the logging code I had to remove the `.date()` except for 
* data_structures_fifo.py::pop()::sell_date
* StockSplitEvent::date

In these cases if I remove the `.date()` function, then it prints the whole datetime.

Reply:
after the updates last week, most timestamps coming from pandas now should immediately be converted to dates, feel free to check again :)